src/charm.py

```diff
@@ -24,22 +28,80 @@ def __init__(self, *args: typing.Any) -> None:
             args: passthrough to CharmBase.
         """
         super().__init__(*args)
+        self._charm_state = CharmState(charm_config=self.config)
+        self._webserver = GunicornWebserver(charm_state=self._charm_state)
         self.framework.observe(self.on.config_changed, self.config_service)
 
+    def flask_container(self, require_connected: bool = True) -> Container:
+        """Get the flask application workload container controller.
+
+        Args:
+            require_connected: if set to ``True``, a runtime exception will be raised if the
+                pebble service inside the container is not ready.
+
+        Return:
+            The controller of the flask application workload container.
+
+        Raises:
+            RuntimeError: if the pebble service inside the container is not ready while the
+                ``require_connected`` is set to True.
+        """
+        container = self.unit.get_container("flask-app")
+        if require_connected and not container.can_connect():
+            raise RuntimeError("pebble inside flask-app container is not ready")
+        return container
+
     def config_service(self, event: ConfigChangedEvent) -> None:
         """Configure the flask pebble service layer.
 
         Args:
             event: the config-changed event that trigger this callback function.
         """
-        container = self.unit.get_container("flask-app")
+        container = self.flask_container(require_connected=False)
         if not container.can_connect():
             event.defer()
             return
+
+        service_name = "flask-app"
         container.add_layer("flask-app", self.flask_layer(), combine=True)
+        webserver_config_path = str(self._webserver.config_path)
+        current_webserver_config = self.pull_file(webserver_config_path)
+        is_webserver_running = container.get_service(service_name).is_running()
+        if current_webserver_config != self._webserver.config:
+            self.push_file(webserver_config_path, self._webserver.config)
+            if is_webserver_running:
+                logger.info("gunicorn config changed, reloading")
+                container.send_signal(self._webserver.reload_signal, service_name)
         container.replan()
         self.unit.status = ActiveStatus()
 
+    def push_file(self, path: str, content: str) -> None:
```

Comment:
I think it will be worth it to move this method and `pull_file` into a dedicated `pebble_helper` module

Reply:
If I would like the move the `pull_file` and `push_file` into a new class, all the pebble APIs (replan, layer, signal) has to be moved to that class too. Then at that time, the new class will be an abstraction over the `ops.model.Container`. But the `ops.model.Container` is already an abstraction over the pebble client. This make the new class an abstraction over an abstraction, which increase the complexity of the code.

The `pull_file` and `push_file` is just here so I can align the pebble file interface with the Python mapping interface (`get` and `__setitem__`) to make the mocking simpler.

Reply:
I agree that right now it's not useful, it will be for a later discussion if the charm.py grows